src/db/mysql/models.rs

```diff
@@ -573,16 +649,56 @@ impl MysqlDb {
         let counts = bso::table
             .select((bso::collection_id, sql::<BigInt>("COUNT(collection_id)")))
             .filter(bso::user_id.eq(user_id.legacy_id as i32))
-            .filter(bso::expiry.gt(&self.session.timestamp))
+            .filter(bso::expiry.gt(&self.timestamp()))
             .group_by(bso::collection_id)
             .load(&self.conn)?
             .into_iter()
             .collect();
         self.map_collection_names(counts)
     }
+
+    pub fn timestamp(&self) -> i64 {
+        self.session.borrow().timestamp
+    }
+
+    #[cfg(test)]
+    pub(super) fn set_timestamp(&self, timestamp: i64) {
+        self.session.borrow_mut().timestamp = timestamp;
+    }
+}
+
+macro_rules! sync_db_method {
+    ($name:ident, $sync_name:ident, $type:ident) => {
+        fn $name(&self, params: params::$type) -> DbFuture<results::$type> {
+            let db = self.clone();
+            Box::new(
+                self.thread_pool
+                    .spawn_handle(lazy(move || future::result(db.$sync_name(params)))),
+            )
+        }
+    }
 }
 
 impl Db for MysqlDb {
+    fn commit(&self) -> DbFuture<()> {
+        let db = self.clone();
+        Box::new(
+            self.thread_pool
+                .spawn_handle(lazy(move || future::result(db.commit_sync()))),
+        )
+    }
```

Comment:
For my own edification, why can't the `sync_db_method!` macro be used with `commit` and `rollback` here?

Reply:
Looks like it's the lack of a `params` argument. Macros get messy pretty quickly when you try to make things optional I've found, because you have to use `*` so you're also introducing repeating groups to the macro body. The simpler the better.